### PR TITLE
BUILD-469: add shared resource webhook image build

### PR DESCRIPTION
**Pull request by jkhelil**
context_dir specifies the subdirectory of the Git repository to be used when building. though it should be set to the hole repository in the case of csi-driver-shared-resource

### CODE REVIEW

File: ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-master.yaml
Comment (gabemontero) [inline ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-master.yaml:28]: Yeah I'm curious if we really need 
```
  inputs:
    ocp_builder_rhel-8-golang-1.17-openshift-4.11:
      as:
      - registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.17-openshift-4.11
```

like we have for the driver image 

let's go without it like you have here, see what happens

**Review by gabemontero — CHANGES_REQUESTED** [review]
update the 4.11 and 4.12 items like you have with master branch

**Comment by gabemontero** [comment]
/retitle BUILD-469: specify the correct dockerfile_path and remove context_dir for csi-driver-shared-resource

**Comment by gabemontero** [comment]
@jkhelil go back and check out my https://github.com/openshift/release/pull/27951#discussion_r853254631 comment

go ahead and update the 4.11 and 4.12 versions as well as the master version

then when the non-rehearsal e2e's are green I'll tag for merge

**Comment by jkhelil** [comment]
> @jkhelil go back and check out my [#27951 (comment)](https://github.com/openshift/release/pull/27951#discussion_r853254631) comment
> 
> go ahead and update the 4.11 and 4.12 versions as well as the master version
> 
> then when the non-rehearsal e2e's are green I'll tag for merge

@gabemontero I didnt see the last comment, sorry about tagging you before that, so now is fixed, i update for 4.11 and 4.12

File: ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-release-4.12.yaml
Comment (gabemontero) [inline ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-release-4.12.yaml:33]: yep double checked and there is no `openshift-4.12` yet so this is correct as I understand it

**Comment by gabemontero** [comment]
/approve

but with adding 4.11 and 4.12 to coincide with master @jkhelil you'll need to run `make openshift-image-mirror-mappings` again and we'll need another approve from @bparees 

see https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/27951/pull-ci-openshift-release-openshift-image-mirror-mappings/1516762329440587776

**Comment by gabemontero** [comment]
/assign @bparees 

we forgot to update the 4.11 and 4.12 config files with the master config file update we did in https://github.com/openshift/release/pull/27887

I had also given @jkhelil bad guidance on use context dir with https://github.com/openshift/release/pull/27887

**Comment by bparees** [comment]
/approve
name looks good

**Comment by bparees** [comment]
/hold

your images job is still failing

**Comment by gabemontero** [comment]
> /hold
> 
> your images job is still failing

yeah @bparees that is going to fail until https://github.com/openshift/csi-driver-shared-resource/pull/92 merges as @jkhelil is moving the Dockerfiles into a subdirectory.

i.e. from the failed images job:

`error: open /tmp/build/inputs/images/Dockerfile: no such file or directory`

Are you OK with our "fixing" things after this merges with https://github.com/openshift/csi-driver-shared-resource/pull/92 , or do you really prefer that @jkhelil need to keep the various Dockerfiles in the repo's top level dir so that this PRs job passes?

I liked the idea of moving the Dockerfiles into a subdir so have been trying to help that approach along, but yeah, it is at the expense of the rehearsals validating things here.

**Comment by bparees** [comment]
> Are you OK with our "fixing" things after this merges with https://github.com/openshift/csi-driver-shared-resource/pull/92 , or do you really prefer that @jkhelil need to keep the various Dockerfiles in the repo's top level dir so that this PRs job passes?

i don't think we should merge a config that can't be built

**Comment by gabemontero** [comment]
> > Are you OK with our "fixing" things after this merges with [openshift/csi-driver-shared-resource#92](https://github.com/openshift/csi-driver-shared-resource/pull/92) , or do you really prefer that @jkhelil need to keep the various Dockerfiles in the repo's top level dir so that this PRs job passes?
> 
> i don't think we should merge a config that can't be built

10-4

/lgtm cancel

@jkhelil I'll post comments in line on what you'll have to change

File: ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-master.yaml
Comment (gabemontero) [inline ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-master.yaml:25]:
```suggestion
- dockerfile_path: Dockerfile.mustgather
```

@jkhelil you'll have to refactor https://github.com/openshift/csi-driver-shared-resource/pull/92 so that the dockerfiles live in the repo's top level dir

File: ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-master.yaml
Comment (gabemontero) [inline ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-master.yaml:18]:
```suggestion
- dockerfile_path: Dockerfile
```

@jkhelil you'll have to refactor https://github.com/openshift/csi-driver-shared-resource/pull/92 so that the dockerfiles live in the repo's top level dir

File: ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-release-4.11.yaml
Comment (gabemontero) [inline ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-release-4.11.yaml:25]:
```suggestion
- dockerfile_path: Dockerfile.mustgather
```

@jkhelil you'll have to refactor https://github.com/openshift/csi-driver-shared-resource/pull/92 so that the dockerfiles live in the repo's top level dir

File: ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-release-4.12.yaml
Comment (gabemontero) [reply ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-release-4.12.yaml:28]: @jkhelil you'll have to refactor https://github.com/openshift/csi-driver-shared-resource/pull/92 so that the dockerfiles live in the repo's top level dir

**Comment by bparees** [comment]
to be clear, the reason why I say that we want the config to be buildable is because if openshift/images can't be built, then i'm not sure what happens when we try to build new payloads if one or more images can't be built successfully.  I realize this image isn't "in the payload" per se, but in general `openshift/images` seems like a job that has to be evergreen to avoid potentially blocking someone.

that means if you want to move the dockerfiles, you'll need to first duplicate them to the new location, then update this config to use the new location, then remove the originals from the old location.

cc @jupierce and @petr-muller who can tell me if i'm being overly paranoid here.

if i'm *not* being overly paranoid, then it seems like `openshift/images` should be a required job on this repo to ensure no one merges something that breaks it (though i realize not all PRs to this repo affect image building)

File: ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-master.yaml
Comment (jkhelil) [reply ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-master.yaml]: this is now fixed

File: ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-master.yaml
Comment (jkhelil) [reply ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-master.yaml]: this now fixed with the correct path

File: ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-master.yaml
Comment (jkhelil) [reply ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-master.yaml]: done with the correct path

File: ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-release-4.11.yaml
Comment (jkhelil) [reply ci-operator/config/openshift/csi-driver-shared-resource/openshift-csi-driver-shared-resource-release-4.11.yaml]: fixed with the correct path, also removing the images subdir from the target PR

**Comment by gabemontero** [comment]
> to be clear, the reason why I say that we want the config to be buildable is because if openshift/images can't be built, then i'm not sure what happens when we try to build new payloads if one or more images can't be built successfully. I realize this image isn't "in the payload" per se, but in general `openshift/images` seems like a job that has to be evergreen to avoid potentially blocking someone.
> 
> that means if you want to move the dockerfiles, you'll need to first duplicate them to the new location, then update this config to use the new location, then remove the originals from the old location.

yep ^^ is the approach we are taking @bparees - thanks 

I was composing https://github.com/openshift/csi-driver-shared-resource/pull/92#pullrequestreview-947526781 while you were posting this :-) 
> 
> cc @jupierce and @petr-muller who can tell me if i'm being overly paranoid here.
> 
> if i'm _not_ being overly paranoid, then it seems like `openshift/images` should be a required job on this repo to ensure no one merges something that breaks it (though i realize not all PRs to this repo affect image building)

**Comment by gabemontero** [comment]
ok now we should get green image runs here after @jkhelil changes in https://github.com/openshift/csi-driver-shared-resource/pull/92 pass e2e's and merge and the new dockerfiles are available

**Comment by gabemontero** [comment]
the recipe to get to green appears to be 
- https://github.com/openshift/release/pull/27998 should simple undo PR #27887 so we can get image builds out of the shared resource repo
- after ^^ merges, https://github.com/openshift/csi-driver-shared-resource/pull/92 should be able to run green with our existing images, but also add the dockerfile for the new webhook image
- after ^^ merges, this PR will need to rebase, re-run `make openshift-image-mirror-mappings`, and should now be able to find all 3 dockerfiles and get a green images job etc.

**Comment by gabemontero** [comment]
ok @jkhelil we can retest after you rebase and I think the images job should pass now that if will be able to find all 3 dockerfiles

**Comment by gabemontero** [comment]
/retitle BUILD-469: add shared resource webhook image build

**Comment by gabemontero** [comment]
and they are all green :-) 

/lgtm

**Pull request merged**
